### PR TITLE
MINOR: Standardize controller log4j output for replaying records

### PR DESCRIPTION
**Pull request by cmccabe**
Standardize controller log4j output for replaying important records. The log message should include word "replayed" to make it clear that this is a record replay. Log the replay of records for ACLs, client quotas, and producer IDs, which were previously not logged. Also fix a case where we weren't logging changes to broker registrations.

AclControlManager, ClientQuotaControlManager, and ProducerIdControlManager didn't previously have a log4j logger object, so this PR adds one. It also converts them to using Builder objects. This makes junit tests more readable because we don't need to specify paramaters where the test can use the default (like LogContexts).

Throw an exception in replay if we get another TopicRecord for a topic which already exists.

Example log messages:
  INFO [QuorumController id=3000] Replayed a FeatureLevelRecord setting metadata version to 3.6-IV0
  DEBUG [QuorumController id=3000] Replayed a ZkMigrationStateRecord which did not alter the state from NONE.
  INFO [QuorumController id=3000] Replayed BrokerRegistrationChangeRecord modifying the registration for broker 0: BrokerRegistrationChangeRecord(brokerId=0, brokerEpoch=3, fenced=-1, inControlledShutdown=0)
  INFO [QuorumController id=3000] Replayed ClientQuotaRecord for ClientQuotaEntity(entries={user=testkit}) setting request_percentage to 0.99.

### CODE REVIEW

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 1]
```diff
@@ -473,8 +488,8 @@ public void replay(PartitionChangeRecord record) {
 
         if (record.removingReplicas() != null || record.addingReplicas() != null) {
             log.info("Replayed partition assignment change {} for topic {}", record, topicInfo.name);
-        } else if (log.isTraceEnabled()) {
-            log.trace("Replayed partition change {} for topic {}", record, topicInfo.name);
+        } else if (log.isDebugEnabled()) {
```
Comment (divijvaidya) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:491]: Do we really need this check? My understanding is that the parameterised messages (which we are using here) removes the requirement of this check. see: https://www.slf4j.org/faq.html#logging_performance

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 2]
```diff
@@ -377,7 +377,19 @@ private ReplicationControlManager(
...
+            // unless, of course, there is a RemoveTopicRecord in between.
+            if (existingUuid.equals(record.topicId())) {
+                throw new RuntimeException("Found duplicate TopicRecord for " + record.name() +
+                        " with topic ID " + record.topicId());
+            } else {
+                throw new RuntimeException("Found duplicate TopicRecord for " + record.name() +
+                        " with a different ID than before. Previous ID was " + existingUuid +
+                        " and new ID is " + record.topicId());
+            }
+        }
```
Comment (divijvaidya) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:392]: 1. This should perhaps go in a separate PR? It would help keeping commits as separate (thus helping in faster rollback if required)
2. I am curious to understand where the RuntimeException gets logged (probably handled by event queue?)

File: metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java [hunk 3]
```diff
@@ -64,12 +66,44 @@
...
+            return this;
+        }
+
+        Builder setClusterMetadataAuthorizer(Optional<ClusterMetadataAuthorizer> authorizer) {
+            this.authorizer = authorizer;
+            return this;
+        }
+
+        AclControlManager build() {
+            if (logContext == null) logContext = new LogContext();
```
Comment (divijvaidya) [inline metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java:90]: Do you think it would be useful to add a LogPrefix here (and other places changed in this PR) such as 
```
LogContext("[AclControlManager id=nodeId")
```

Comment (rondagostino) [reply metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java:90]: I agree this is a worthwhile question to consider.  I do note that the builders that already exist don't set anything special.  For example, from `ConfigurationControlManager.Builder.build()`:

```
            if (logContext == null) logContext = new LogContext();
```

And all of these builder log contexts do get explicitly set in QuorumController (including the new ones via this PR), so the default will never be used.

So my guess is the answer is no, it isn't worth doing.  Colin, WDYT?

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 4]
```diff
@@ -377,7 +377,19 @@ private ReplicationControlManager(
...
 
     public void replay(TopicRecord record) {
-        topicsByName.put(record.name(), record.topicId());
+        Uuid existingUuid = topicsByName.put(record.name(), record.topicId());
+        if (existingUuid != null) {
+            // We don't currently support sending a second TopicRecord for the same topic name...
+            // unless, of course, there is a RemoveTopicRecord in between.
+            if (existingUuid.equals(record.topicId())) {
+                throw new RuntimeException("Found duplicate TopicRecord for " + record.name() +
+                        " with topic ID " + record.topicId());
```
Comment (rondagostino) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:386]: nit: Noting that this first clause is explicitly rejecting the possibility of an idempotent replay of the same record a second time.  This makes sense, as it is unexpected if it should happen. 
 It might be good to call this out explicitly in a comment.

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 1 above)
Comment (rondagostino) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:491]: +1

File: metadata/src/test/java/org/apache/kafka/controller/MockAclControlManager.java [hunk 5]
```diff
@@ -33,7 +33,7 @@
 public class MockAclControlManager extends AclControlManager {
     public MockAclControlManager(LogContext logContext,
                                  Optional<ClusterMetadataAuthorizer> authorizer) {
-        super(new SnapshotRegistry(logContext), authorizer);
+        super(new LogContext(), new SnapshotRegistry(logContext), authorizer);
```
Comment (rondagostino) [inline metadata/src/test/java/org/apache/kafka/controller/MockAclControlManager.java:36]: Seems like it should be `s/new LogContext()/logContext/`?

**Review by rondagostino — COMMENTED** [review]
LGTM in general, just a few minor comments and a couple of places where it would be good to get your take on the issue raised, Colin.

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 1 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:491]: good point. I'll remove it.

File: metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java (hunk 3 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java:90]: @rondagostino is correct that the defaults here do not get used except in unit tests.

In general we want `[QuorumController id=1] ` to appear before every quorum controller log message. The class name (AclControlManager or whatever) actually gets logged separately at the end of the line, when using the typical log4j setup.

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 4 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:386]: The comment says:
```
            // We don't currently support sending a second TopicRecord for the same topic name...
            // unless, of course, there is a RemoveTopicRecord in between.
```

which I think does cover the scenario you were thinking of (let me know if this makes sense)

File: metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java (hunk 3 above)
Comment (divijvaidya) [reply metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java:90]: ok, sounds good.

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 2 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:392]: The RuntimeException is handled in `handleEventException`

File: metadata/src/test/java/org/apache/kafka/controller/MockAclControlManager.java (hunk 5 above)
Comment (cmccabe) [reply metadata/src/test/java/org/apache/kafka/controller/MockAclControlManager.java:36]: fixed, thanks.

**Comment by cmccabe** [comment]
> And all of these builder log contexts do get explicitly set in QuorumController (including the new ones via this PR), so the default will never be used. So my guess is the answer is no, it isn't worth doing. Colin, WDYT?

Well, the JUnit tests don't set the log context. The builder is useful for avoiding clutter in these tests.

In general most of the controller manager objects should have builders. It just avoids having a huge amount of boilerplate in the tests (like dummy LogContext, SnapshotRegistry, objects etc.) It also makes it bearable to add new constructor parameters without updating every since test call site.

**Review by mumrah — APPROVED** [review]
Thanks @cmccabe, this is a nice improvement. It's really confusing to look at controller logs and see nodes saying "Creating topic..." when really its just replaying the record.

Can you paste a snippet of some INFO logs with these changes and include in the PR description?

LGTM

**Comment by jolshan** [comment]
Guys -- we really need to check builds before merging. Every time this happens, we are blocking people who are doing the right thing and checking their builds :(

**Comment by ijuma** [comment]
+1 @jolshan - is there a reason why people are not checking the build result? Particularly when it doesn't even compile.

**Comment by cmccabe** [comment]
As I recall, we did check the Jenkins result, it's just that it was an old and stale build result. I should have re-run the build. Sorry.
